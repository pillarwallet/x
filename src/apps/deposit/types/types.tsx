export type BalanceInfo = {
  type?: 'BalanceInfo';
  name: string;
  chain: string;
  address: string;
  decimals: number;
  symbol: string;
  logoURI: string;
  balance: string;
};

export type AddedAssets = {
  type?: 'AddedAsset';
  assetType: 'token' | 'nft';
  tokenId?: string;
  chain: string;
  tokenAddress: string;
  balance: number;
};

export type Network =
  | 'ethereum'
  | 'polygon'
  | 'gnosis'
  | 'base'
  | 'bnb smart chain'
  | 'optimism'
  | 'arbitrum';

export type TokenList = {
  address: string;
  chainId: number;
  decimals: string;
  name: string;
  symbol: string;
  logoURI: string;
};

export type BalancesByChain = {
  [chain: string]: BalanceInfo[];
};
