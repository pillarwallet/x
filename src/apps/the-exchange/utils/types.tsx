import { BridgingProvider } from '@etherspot/data-utils/dist/cjs/sdk/data/constants';
import { Route } from '@lifi/sdk';
import { Hex } from 'viem';

export enum CardPosition {
  SWAP = 'SWAP',
  RECEIVE = 'RECEIVE',
}

export type SwapType = {
  fromAmount: number;
  fromTokenAddress: string;
  fromChainId: number;
  fromTokenDecimals: number;
  toTokenAddress: string;
  toChainId: number;
  toTokenDecimals: number;
  slippage: number;
  fromAccountAddress?: string;
  provider?: BridgingProvider;
};

export type SwapOffer = {
  tokenAmountToReceive: number;
  offer: Route;
};

export type ChainType = {
  chainId: number;
  chainName: string;
};

export type StepTransaction = {
  to?: string;
  data?: Hex;
  value?: bigint;
  gasLimit?: bigint;
  gasPrice?: bigint;
  chainId?: number;
  type?: number | string;
  transactionType?: StepType;
};

export type StepType = 'swap' | 'cross' | 'lifi' | 'custom' | 'approval';
