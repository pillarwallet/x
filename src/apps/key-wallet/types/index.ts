export interface Asset {
  id: number;
  name: string;
  symbol: string;
  logo: string;
  balance: number;
  decimals: number;
  price: number;
  price_change_24h: number;
  contract: string;
  chainId: number;
  chainName: string;
  usdBalance: number;
}

export interface SendFormData {
  amount: string;
  recipient: string;
}

export interface TransactionStatus {
  hash: string;
  chainId: number;
  status: 'pending' | 'success' | 'failed';
  timestamp: number;
}

