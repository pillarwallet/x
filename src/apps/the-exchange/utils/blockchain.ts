import { BigNumber, BigNumberish } from 'ethers';
import { formatEther, formatUnits } from 'ethers/lib/utils';
import { decodeFunctionData, erc20Abi } from 'viem';

// types
import { Token } from '../../../services/tokensData';
import { StepTransaction } from './types';

export const processBigNumber = (val: BigNumber): number =>
  Number(val.toString());

export const processEth = (val: BigNumberish, dec: number): number => {
  if (typeof val === 'bigint') {
    return +parseFloat(formatEther(val)).toFixed(2);
  }

  return +parseFloat(formatUnits(val as BigNumberish, dec));
};

// Utility: get native token symbol for a chain
export const NATIVE_SYMBOLS: Record<number, string> = {
  1: 'ETH',
  100: 'xDAI',
  137: 'POL',
  10: 'ETH',
  42161: 'ETH',
  56: 'BNB',
  8453: 'ETH',
};

// Helper: Detect if a tx is a native fee step
export const isNativeFeeTx = (
  tx: StepTransaction,
  feeReceiver: string
): boolean => {
  return (
    typeof tx.to === 'string' &&
    typeof feeReceiver === 'string' &&
    tx.to.toLowerCase() === feeReceiver.toLowerCase()
  );
};

// Helper: Detect if a tx is an ERC20 (stablecoin or wrapped) fee step
export const isERC20FeeTx = (
  tx: StepTransaction,
  swapToken: Token
): boolean => {
  return (
    typeof tx.to === 'string' &&
    typeof swapToken.contract === 'string' &&
    tx.to.toLowerCase() === swapToken.contract.toLowerCase() &&
    tx.value === BigInt(0) &&
    typeof tx.data === 'string' &&
    tx.data.startsWith('0xa9059cbb')
  );
};

// Helper: Extract fee amount from tx
export const getFeeAmount = (
  tx: StepTransaction,
  swapToken: Token,
  decimals: number
): string => {
  if (tx.value && tx.data === '0x') {
    // Native
    return formatEther(tx.value);
  }
  if (isERC20FeeTx(tx, swapToken)) {
    try {
      const decoded = decodeFunctionData({
        abi: erc20Abi,
        data: tx.data || '0x',
      });
      if (decoded.args && typeof decoded.args[1] === 'bigint') {
        return formatUnits(decoded.args[1], decimals);
      }
    } catch (e) {
      return '0';
    }
  }
  return '0';
};

// Helper: Get fee symbol
export const getFeeSymbol = (
  tx: StepTransaction,
  swapToken: Token,
  chainId: number
): string => {
  if (tx.value && tx.data === '0x') {
    return NATIVE_SYMBOLS[chainId] || 'NATIVE';
  }
  return swapToken.symbol;
};
