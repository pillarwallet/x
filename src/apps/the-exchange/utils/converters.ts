export const hasThreeZerosAfterDecimal = (num: number): boolean => {
  const decimalPart = num.toString().split('.')[1] || '';
  return decimalPart.startsWith('000');
};

export const formatTokenAmount = (amount?: number) => {
  if (amount === undefined) return 0;
  return hasThreeZerosAfterDecimal(amount)
    ? amount.toFixed(8)
    : amount.toFixed(4);
};
