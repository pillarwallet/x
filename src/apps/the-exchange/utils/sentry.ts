import * as Sentry from '@sentry/react';
import { useWalletAddress } from '@etherspot/transaction-kit';

// Sentry configuration for the-exchange app
export const initSentryForExchange = () => {
  Sentry.setTag('app', 'the-exchange');
  Sentry.setTag('module', 'exchange');
};

// Utility to get fallback wallet address for logging
// This function should be called from within a React component context
// where the wallet address is available
export const fallbackWalletAddressForLogging = (): string => {
  // This is a utility function that should be called from within React components
  // The actual wallet address should be passed as a parameter or obtained via hook
  // For now, return unknown as this is a fallback utility
  return 'unknown_wallet_address';
};

// Enhanced Sentry logging with wallet address context
export const logExchangeEvent = (
  message: string,
  level: Sentry.SeverityLevel = 'info',
  extra?: Record<string, unknown>,
  tags?: Record<string, string>
) => {
  const walletAddress = fallbackWalletAddressForLogging();

  Sentry.withScope((scope) => {
    scope.setLevel(level);
    scope.setTag('wallet_address', walletAddress);
    scope.setTag('app_module', 'the-exchange');

    if (tags) {
      Object.entries(tags).forEach(([key, value]) => {
        scope.setTag(key, value);
      });
    }

    if (extra) {
      scope.setExtra('exchange_data', extra);
    }

    Sentry.captureMessage(message);
  });
};

// Log exchange errors with wallet address
export const logExchangeError = (
  error: Error | string,
  extra?: Record<string, unknown>,
  tags?: Record<string, string>
) => {
  const walletAddress = fallbackWalletAddressForLogging();

  Sentry.withScope((scope) => {
    scope.setLevel('error');
    scope.setTag('wallet_address', walletAddress);
    scope.setTag('app_module', 'the-exchange');
    scope.setTag('error_type', 'exchange_error');

    if (tags) {
      Object.entries(tags).forEach(([key, value]) => {
        scope.setTag(key, value);
      });
    }

    if (extra) {
      scope.setExtra('exchange_error_data', extra);
    }

    if (error instanceof Error) {
      Sentry.captureException(error);
    } else {
      Sentry.captureMessage(error, 'error');
    }
  });
};

// Generic operation logging function
export const logOperation = (
  operationType: string,
  operation: string,
  data: Record<string, unknown>,
  walletAddress?: string
) => {
  Sentry.withScope((scope) => {
    scope.setLevel('info');
    scope.setTag(
      'wallet_address',
      walletAddress || fallbackWalletAddressForLogging()
    );
    scope.setTag('app_module', 'the-exchange');
    scope.setTag('operation_type', operationType);
    scope.setTag(`${operationType}_operation`, operation);

    scope.setExtra(`${operationType}_data`, data);

    Sentry.captureMessage(
      `${operationType.charAt(0).toUpperCase() + operationType.slice(1)} operation: ${operation}`,
      'info'
    );
  });
};

// Log swap operations
export const logSwapOperation = (
  operation: string,
  data: Record<string, unknown>,
  walletAddress?: string
) => {
  logOperation('swap', operation, data, walletAddress);
};

// Log token operations
export const logTokenOperation = (
  operation: string,
  tokenData: Record<string, unknown>,
  walletAddress?: string
) => {
  logOperation('token', operation, tokenData, walletAddress);
};

// Log offer operations
export const logOfferOperation = (
  operation: string,
  offerData: Record<string, unknown>,
  walletAddress?: string
) => {
  logOperation('offer', operation, offerData, walletAddress);
};

// Log transaction operations
export const logTransactionOperation = (
  operation: string,
  transactionData: Record<string, unknown>,
  walletAddress?: string
) => {
  logOperation('transaction', operation, transactionData, walletAddress);
};

// Log user interactions
export const logUserInteraction = (
  interaction: string,
  data: Record<string, unknown>,
  walletAddress?: string
) => {
  Sentry.withScope((scope) => {
    scope.setLevel('info');
    scope.setTag(
      'wallet_address',
      walletAddress || fallbackWalletAddressForLogging()
    );
    scope.setTag('app_module', 'the-exchange');
    scope.setTag('interaction_type', 'user_action');
    scope.setTag('user_interaction', interaction);

    scope.setExtra('interaction_data', data);

    Sentry.captureMessage(`User interaction: ${interaction}`, 'info');
  });
};

// Log performance metrics
export const logPerformanceMetric = (
  metric: string,
  value: number,
  unit: string = 'ms',
  walletAddress?: string
) => {
  Sentry.withScope((scope) => {
    scope.setLevel('info');
    scope.setTag(
      'wallet_address',
      walletAddress || fallbackWalletAddressForLogging()
    );
    scope.setTag('app_module', 'the-exchange');
    scope.setTag('metric_type', 'performance');
    scope.setTag('metric_name', metric);

    scope.setExtra('performance_data', {
      metric,
      value,
      unit,
      timestamp: new Date().toISOString(),
    });

    Sentry.captureMessage(
      `Performance metric: ${metric} = ${value}${unit}`,
      'info'
    );
  });
};

// Hook to get wallet address for logging
export const useWalletAddressForLogging = () => {
  const walletAddress = useWalletAddress();
  return walletAddress || 'unknown_wallet_address';
};

// Note: Error boundary removed due to TypeScript/JSX compatibility issues
// Use Sentry's default error boundary or create a separate React component file

// Transaction monitoring for exchange operations
export const startExchangeTransaction = (
  operation: string,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  _data: Record<string, unknown> = {},
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  _walletAddress?: string
) => {
  return Sentry.startSpan(
    {
      name: `exchange.${operation}`,
      op: 'exchange.operation',
    },
    (span) => {
      // Note: Span API has changed in Sentry v10
      // Properties are set via the span context instead
      return span;
    }
  );
};

// Breadcrumb utilities for exchange
export const addExchangeBreadcrumb = (
  message: string,
  category: string = 'exchange',
  data?: Record<string, unknown>,
  level: Sentry.SeverityLevel = 'info'
) => {
  const walletAddress = fallbackWalletAddressForLogging();

  Sentry.addBreadcrumb({
    message,
    category,
    level,
    data: {
      ...data,
      wallet_address: walletAddress,
      app_module: 'the-exchange',
    },
  });
};
