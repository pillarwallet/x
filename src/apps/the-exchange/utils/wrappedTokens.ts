export const NATIVE_TOKEN_ADDRESSES = new Set([
  '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee',
  '0x0000000000000000000000000000000000000000',
]);

// Not including XDAI below
export const WRAPPED_NATIVE_TOKEN_ADDRESSES: Record<number, string> = {
  // Ethereum
  1: '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2', // WETH
  // Polygon
  137: '0x0d500B1d8E8eF31E21C99d1Db9A6444d3ADf1270', // WMATIC
  // Optimism
  10: '0x4200000000000000000000000000000000000006', // WETH
  // Arbitrum
  42161: '0x82aF49447D8a07e3bd95BD0d56f35241523fBab1', // WETH
  // Base
  8453: '0x4200000000000000000000000000000000000006', // WETH
  // BNB
  56: '0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c', // WBNB
};

export const isWrappedToken = (tokenAddress: string, chainId: number) => {
  return (
    tokenAddress.toLowerCase() ===
    WRAPPED_NATIVE_TOKEN_ADDRESSES[chainId]?.toLowerCase()
  );
};

export const isNativeToken = (address: string) =>
  NATIVE_TOKEN_ADDRESSES.has(address.toLowerCase());

export const getWrappedTokenAddressIfNative = (
  tokenAddress: string,
  chainId: number
): string => {
  if (isNativeToken(tokenAddress)) {
    const wrappedAddress = WRAPPED_NATIVE_TOKEN_ADDRESSES[chainId];
    // fallback to token address in the case of XDAI token
    return wrappedAddress || tokenAddress;
  }
  return tokenAddress;
};
