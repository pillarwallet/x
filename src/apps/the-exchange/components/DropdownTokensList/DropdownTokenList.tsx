import { Token } from '@etherspot/data-utils/dist/cjs/sdk/data/classes/token';
import { useWalletAddress } from '@etherspot/transaction-kit';
import { useState } from 'react';

// api
import { useRecordPresenceMutation } from '../../../../services/pillarXApiPresence';

// reducer
import {
  setIsReceiveOpen,
  setIsSwapOpen,
  setReceiveChain,
  setReceiveToken,
  setSearchTokenResult,
  setSwapChain,
  setSwapToken,
} from '../../reducer/theExchangeSlice';

// hooks
import { useAppDispatch, useAppSelector } from '../../hooks/useReducerHooks';

// types
import { CardPosition, ChainType } from '../../utils/types';

// utils
import { convertChainIdtoName } from '../../../../utils/blockchain';

// components
import SelectDropdown from '../SelectDropdown/SelectDropdown';
import TokenListItem from '../TokenListItem/TokenListItem';
import TokenSearchInput from '../TokenSearchInput/TokenSearchInput';

// images
import CloseIcon from '../../images/add.png';

type DropdownTokenListProps = {
  type: CardPosition;
  initialCardPosition: CardPosition;
};

const DropdownTokenList = ({
  type,
  initialCardPosition,
}: DropdownTokenListProps) => {
  /**
   * Import the recordPresence mutation from the
   * pillarXApiPresence service. We use this to
   * collect data on what asset is being selected
   */
  const [recordPresence] = useRecordPresenceMutation();

  const accountAddress = useWalletAddress();

  const dispatch = useAppDispatch();
  const isSwapOpen = useAppSelector(
    (state) => state.swap.isSwapOpen as boolean
  );
  const swapTokenData = useAppSelector(
    (state) => state.swap.swapTokenData as Token[]
  );
  const receiveTokenData = useAppSelector(
    (state) => state.swap.receiveTokenData as Token[]
  );
  const searchTokenResult = useAppSelector(
    (state) => state.swap.searchTokenResult as Token[]
  );
  const swapChain = useAppSelector(
    (state) => state.swap.swapChain as ChainType
  );
  const receiveChain = useAppSelector(
    (state) => state.swap.receiveChain as ChainType
  );
  const swapToken = useAppSelector((state) => state.swap.swapToken as Token);
  const receiveToken = useAppSelector(
    (state) => state.swap.receiveToken as Token
  );

  const [isChainSelectionOpen, setIsChainSelectionOpen] =
    useState<boolean>(false);

  // select all chainsId of tokens available in the list for swap token
  const allChainsSwap = swapTokenData?.map((chain) => chain.chainId);
  const uniqueChainsSwap = allChainsSwap.filter((chain, index) => {
    return allChainsSwap.indexOf(chain) === index;
  });

  // select all chainsId of tokens available in the list for receive token
  const allChainsReceive = receiveTokenData?.map((chain) => chain.chainId);
  const uniqueChainsReceive = allChainsReceive.filter((chain, index) => {
    return allChainsReceive.indexOf(chain) === index;
  });

  // if there are no tokens being typed searched, we show the swapTokenData list of tokens
  // which will filter if a chain has been chosen
  let swapTokenList;

  if (searchTokenResult?.length) {
    swapTokenList = searchTokenResult;
  } else if (swapChain?.chainId) {
    swapTokenList = swapTokenData.filter(
      (token) => token.chainId === swapChain.chainId
    );
  } else {
    swapTokenList = swapTokenData;
  }

  // if there are no tokens being typed searched, we show the receiveTokenData list of tokens
  // which will filter if a chain has been chosen
  let receiveTokenList;

  if (searchTokenResult?.length) {
    receiveTokenList = searchTokenResult;
  } else if (receiveChain?.chainId) {
    receiveTokenList = receiveTokenData.filter(
      (token) => token.chainId === receiveChain.chainId
    );
  } else {
    receiveTokenList = receiveTokenData;
  }

  return (
    <>
      <div
        id="dropdown-token-list-exchange"
        className="fixed inset-0 bg-black_grey/[.9] -z-10"
        data-testid="dropdown-token-list"
      >
        <button
          type="button"
          onClick={() => {
            if (type === CardPosition.SWAP) {
              dispatch(setIsSwapOpen(false));
            } else {
              dispatch(setIsReceiveOpen(false));
            }
            dispatch(setSearchTokenResult([]));
          }}
          className="fixed top-0 right-0 w-[50px] h-[50px] mt-6 mr-4 mb-20 desktop:mr-14 desktop:mb-28 bg-black"
          data-testid="close-card-button"
        >
          <img
            src={CloseIcon}
            alt="close-modal-button"
            className="w-full h-auto"
          />
        </button>
      </div>
      <div className="flex flex-col w-full max-w-[420px]">
        <div
          className={`flex flex-row gap-[10px] p-4 w-full rounded-t-[3px] border-b border-b-black_grey ${initialCardPosition === CardPosition.SWAP ? 'bg-light_green' : 'bg-purple'}`}
        >
          <SelectDropdown
            options={isSwapOpen ? uniqueChainsSwap : uniqueChainsReceive}
            isOpen={isChainSelectionOpen}
            onClick={() => setIsChainSelectionOpen(!isChainSelectionOpen)}
            className={`${isChainSelectionOpen && 'w-full'}`}
            onSelect={() => setIsChainSelectionOpen(false)}
          />
          <TokenSearchInput
            placeholder="Search tokens"
            isShrinked={isChainSelectionOpen}
          />
        </div>
        <div
          id="token-list-exchange"
          className={`flex flex-col p-4 w-full rounded-b-[3px] max-h-[272px] mr-4 overflow-y-auto ${initialCardPosition === CardPosition.SWAP ? 'bg-light_green' : 'bg-purple'}`}
        >
          {isSwapOpen
            ? swapTokenList
                .filter(
                  (token) =>
                    token.chainId !== receiveToken?.chainId ||
                    token.address !== receiveToken?.address
                )
                .map((token, index) => (
                  <TokenListItem
                    key={index}
                    testId={`token-list-item-exchange-send-${token.chainId}-${token.name}`}
                    onClick={() => {
                      dispatch(setSwapToken(token));
                      dispatch(
                        setSwapChain({
                          chainId: token.chainId,
                          chainName: convertChainIdtoName(token.chainId),
                        })
                      );
                      recordPresence({
                        address: accountAddress,
                        action: 'app:theExchange:sourceTokenSelect',
                        value: {
                          chainId: token.chainId,
                          address: token.address,
                          symbol: token.symbol,
                          name: token.name,
                        },
                      });
                      dispatch(setSearchTokenResult([]));
                      dispatch(setIsSwapOpen(false));
                    }}
                    tokenName={token.name}
                    tokenSymbol={token.symbol}
                    chainName={convertChainIdtoName(token.chainId)}
                    tokenLogo={token.icon}
                  />
                ))
            : receiveTokenList
                .filter(
                  (token) =>
                    token.chainId !== swapToken?.chainId ||
                    token.address !== swapToken?.address
                )
                .map((token, index) => (
                  <TokenListItem
                    key={index}
                    testId={`token-list-item-exchange-receive-${token.chainId}-${token.name}`}
                    onClick={() => {
                      dispatch(setReceiveToken(token));
                      dispatch(
                        setReceiveChain({
                          chainId: token.chainId,
                          chainName: convertChainIdtoName(token.chainId),
                        })
                      );
                      recordPresence({
                        address: accountAddress,
                        action: 'app:theExchange:destinationTokenSelect',
                        value: {
                          chainId: token.chainId,
                          address: token.address,
                          symbol: token.symbol,
                          name: token.name,
                        },
                      });
                      dispatch(setSearchTokenResult([]));
                      dispatch(setIsReceiveOpen(false));
                    }}
                    tokenName={token.name}
                    tokenSymbol={token.symbol}
                    chainName={convertChainIdtoName(token.chainId)}
                    tokenLogo={token.icon}
                  />
                ))}
        </div>
      </div>
    </>
  );
};

export default DropdownTokenList;
