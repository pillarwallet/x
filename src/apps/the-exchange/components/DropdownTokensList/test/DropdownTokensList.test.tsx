import { fireEvent, render } from '@testing-library/react';
import renderer, { act } from 'react-test-renderer';

// provider
import { Provider } from 'react-redux';
import { store } from '../../../../../store';

// reducer
import {
  setAmountReceive,
  setAmountSwap,
  setBestOffer,
  setIsOfferLoading,
  setIsReceiveOpen,
  setIsSwapOpen,
  setReceiveChain,
  setReceiveToken,
  setReceiveTokenData,
  setSearchTokenResult,
  setSwapChain,
  setSwapToken,
  setSwapTokenData,
  setUsdPriceReceiveToken,
  setUsdPriceSwapToken,
} from '../../../reducer/theExchangeSlice';

// types
import { CardPosition } from '../../../utils/types';

// components
import DropdownTokenList from '../DropdownTokenList';

const mockTokenAssets = [
  {
    address: '0x01',
    name: 'Ether',
    symbol: 'ETH',
    chainId: 1,
    decimals: 18,
    icon: 'iconEth.png',
  },
  {
    address: '0x02',
    name: 'POL',
    symbol: 'POL',
    chainId: 137,
    decimals: 18,
    icon: 'iconMatic.png',
  },
];

describe('<DropdownTokenList />', () => {
  beforeEach(() => {
    act(() => {
      store.dispatch(setSwapTokenData(mockTokenAssets));
      store.dispatch(setReceiveTokenData(mockTokenAssets));
      store.dispatch(setIsSwapOpen(false));
      store.dispatch(setIsReceiveOpen(false));
      store.dispatch(setSwapChain({ chainId: 1, chainName: 'Ethereum' }));
      store.dispatch(setReceiveChain({ chainId: 137, chainName: 'Polygon' }));
      store.dispatch(setSwapToken(mockTokenAssets[0]));
      store.dispatch(setReceiveToken(mockTokenAssets[1]));
      store.dispatch(setAmountSwap(0.1));
      store.dispatch(setAmountReceive(10));
      store.dispatch(setBestOffer(undefined));
      store.dispatch(setSearchTokenResult([]));
      store.dispatch(setUsdPriceSwapToken(1200));
      store.dispatch(setUsdPriceReceiveToken(0.4));
      store.dispatch(setIsOfferLoading(false));
    });
  });
  it('renders correctly and matches snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <DropdownTokenList
            type={CardPosition.SWAP}
            initialCardPosition={CardPosition.SWAP}
          />
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders with initial props', () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <DropdownTokenList
          type={CardPosition.SWAP}
          initialCardPosition={CardPosition.SWAP}
        />
      </Provider>
    );

    act(() => {
      store.dispatch(setIsSwapOpen(true));
    });

    const dropdownList = getByTestId('dropdown-token-list');
    expect(dropdownList).toBeInTheDocument();
  });

  it('allows selecting tokens in Swap card', async () => {
    const { getByPlaceholderText, getAllByTestId } = render(
      <Provider store={store}>
        <DropdownTokenList
          type={CardPosition.SWAP}
          initialCardPosition={CardPosition.SWAP}
        />
      </Provider>
    );

    act(() => {
      store.dispatch(setIsSwapOpen(true));
      store.dispatch(setSwapChain({ chainId: 0, chainName: 'all' }));
    });

    const searchInput = getByPlaceholderText('Search tokens');
    fireEvent.focus(searchInput);
    fireEvent.change(searchInput, { target: { value: 'eth' } });

    const tokenItems = getAllByTestId('token-list-item');
    expect(tokenItems.length).toBe(1);

    const firstTokenItem = getAllByTestId('token-list-item');
    fireEvent.click(firstTokenItem[0]);

    expect(store.getState().swap.swapToken).toBe(mockTokenAssets[0]);
    expect(store.getState().swap.swapChain).toEqual({
      chainId: 1,
      chainName: '1',
    });
    expect(store.getState().swap.isSwapOpen).toBe(false);
  });

  it('allows selecting tokens in Receive card', async () => {
    const { getByPlaceholderText, getAllByTestId } = render(
      <Provider store={store}>
        <DropdownTokenList
          type={CardPosition.RECEIVE}
          initialCardPosition={CardPosition.RECEIVE}
        />
      </Provider>
    );

    act(() => {
      store.dispatch(setIsReceiveOpen(true));
      store.dispatch(setReceiveChain({ chainId: 0, chainName: 'all' }));
    });

    const searchInput = getByPlaceholderText('Search tokens');
    fireEvent.focus(searchInput);
    fireEvent.change(searchInput, { target: { value: 'ma' } });

    const tokenItems = getAllByTestId('token-list-item');
    expect(tokenItems.length).toBe(1);

    const firstTokenItem = getAllByTestId('token-list-item');
    fireEvent.click(firstTokenItem[0]);

    expect(store.getState().swap.receiveToken).toBe(mockTokenAssets[1]);
    expect(store.getState().swap.receiveChain).toEqual({
      chainId: 137,
      chainName: '137',
    });
    expect(store.getState().swap.isReceiveOpen).toBe(false);
  });

  it('closes the dropdown when close button is clicked', async () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <DropdownTokenList
          type={CardPosition.SWAP}
          initialCardPosition={CardPosition.SWAP}
        />
      </Provider>
    );

    act(() => {
      store.dispatch(setIsSwapOpen(true));
    });

    const closeButton = getByTestId('close-card-button');
    fireEvent.click(closeButton);

    expect(store.getState().swap.isSwapOpen).toBe(false);
    expect(store.getState().swap.searchTokenResult).toEqual([]);
  });
});
