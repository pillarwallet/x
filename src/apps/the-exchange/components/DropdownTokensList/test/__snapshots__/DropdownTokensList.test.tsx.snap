// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownTokenList /> renders correctly and matches snapshot 1`] = `
[
  <div
    className="fixed inset-0 bg-black_grey/[.9] -z-10"
    data-testid="dropdown-token-list"
    id="dropdown-token-list-exchange"
  >
    <button
      className="fixed top-0 right-0 w-[50px] h-[50px] mt-6 mr-4 mb-20 desktop:mr-14 desktop:mb-28 bg-black"
      data-testid="close-card-button"
      onClick={[Function]}
      type="button"
    >
      <img
        alt="close-modal-button"
        className="w-full h-auto"
        src="add.png"
      />
    </button>
  </div>,
  <div
    className="flex flex-col w-full max-w-[420px]"
  >
    <div
      className="flex flex-row gap-[10px] p-4 w-full rounded-t-[3px] border-b border-b-black_grey bg-light_green"
    >
      <div
        className="false h-[34px] z-20"
        id="select-chain-dropdown-list-exchange"
      >
        <button
          className="flex justify-between w-full h-full p-2 bg-white rounded-[3px] focus:outline-none"
          onClick={[Function]}
          type="button"
        >
          <p
            className="text-base font-medium text-black_grey text-black capitalize"
          >
            Polygon
          </p>
          <img
            alt="arrow-down"
            src="arrow-down.png"
          />
        </button>
      </div>
      <div
        className="undefined w-full relative h-[34px]"
        id="token-search-input-exchange"
      >
        <input
          className="w-full h-full rounded-[3px] p-2 pr-9 bg-white focus:outline-none focus:ring-0 font-normal text-black text-md placeholder-[#717171] undefined"
          onChange={[Function]}
          placeholder="Search tokens"
          value=""
        />
        <span
          className="absolute inset-y-0 right-0 flex items-center pr-2 cursor-pointer"
          onClick={[Function]}
        >
          <img
            alt="search-icon"
            src="search-icon.png"
          />
        </span>
      </div>
    </div>
    <div
      className="flex flex-col p-4 w-full rounded-b-[3px] max-h-[272px] mr-4 overflow-y-auto bg-light_green"
      id="token-list-exchange"
    >
      <div
        className="flex justify-between items-end border-b py-4 border-b-black_grey group group-hover:border-b-black_grey/[.6] cursor-pointer"
        data-testid="token-list-item"
        id="token-list-item-exchange"
        onClick={[Function]}
      >
        <div
          className="flex items-center"
        >
          <div
            className="flex min-w-fit items-end"
          >
            <img
              alt="token-logo"
              className="w-[30px] h-[30px] object-fill rounded-full grayscale"
              src="iconMatic.png"
            />
            <div
              className="w-3.5 h-3.5 rounded-full border border-white relative right-[7px] hidden"
            >
              <img
                alt="chain-logo"
                className="w-full h-full object-fill rounded-full grayscale hidden"
              />
            </div>
          </div>
          <div
            className="flex flex-col ml-[10px]"
          >
            <p
              className="text-base font-medium text-black_grey group-hover:opacity-60"
            >
              POL
            </p>
            <p
              className="text-xs font-normal text-black_grey group-hover:opacity-60"
            >
              POL
            </p>
          </div>
        </div>
        <p
          className="text-xs font-normal text-black_grey group-hover:opacity-60 capitalize"
        >
          On 
          137
        </p>
      </div>
    </div>
  </div>,
]
`;
